### PR TITLE
Introduce new version v1beta2 for DynaKube and add conversion logic

### PR DESCRIPTION
**Pull request by waodim**
With this introduction and conversion logic of version v1beta2 for the DynaKube is done. Version v1alpha1 is removed for DynaKube and the new hubVersion is v1beta1. It is also the storageVersion

Summary of things that are added:

- a new field metaDataEnrichment was added to the .spec of the Dynakube. it has an enabled field (enabled by default) and a namespace selector.
- a new field dynatraceApiRequestsThreshold was added to the .spec of the Dynakube.
- a new field secCompProfile was added to relevant oneAgent sections: hostMonitoring, classicFullStack, cloudNativeFullstack

Please have a look at the tags if they are correct and also check my wording please.

Corresponding JIRA tickets: [K8S-9331](https://dt-rnd.atlassian.net/browse/K8S-9331) and [K8S-9492](https://dt-rnd.atlassian.net/browse/K8S-9492)
## How can this be tested?

Take your time to:
- deploy operator
- try to query dynakubes
- test basic functionalities

### CODE REVIEW

File: pkg/api/v1beta2/dynakube/feature_flags.go [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+	AnnotationFeatureActiveGateAppArmor                   = AnnotationFeaturePrefix + "activegate-apparmor"
+	AnnotationFeatureAutomaticK8sApiMonitoring            = AnnotationFeaturePrefix + "automatic-kubernetes-api-monitoring"
+	AnnotationFeatureAutomaticK8sApiMonitoringClusterName = AnnotationFeaturePrefix + "automatic-kubernetes-api-monitoring-cluster-name"
+	AnnotationFeatureK8sAppEnabled                        = AnnotationFeaturePrefix + "k8s-app-enabled"
+	AnnotationFeatureActiveGateIgnoreProxy                = AnnotationFeaturePrefix + "activegate-ignore-proxy"
+
+	// dtClient.
+
+	AnnotationFeatureNoProxy             = AnnotationFeaturePrefix + "no-proxy"
+	AnnotationFeatureApiRequestThreshold = AnnotationFeaturePrefix + "dynatrace-api-request-threshold"
```
Comment (0sewa0) [inline pkg/api/v1beta2/dynakube/feature_flags.go:50]: if we have a specific field for it then we shouldn't need it here, the conversion should handle it. (if it was present on the previous dynakube as a feature-flag, then it was moved to the specific field by the conversion)

File: pkg/api/v1beta2/dynakube/feature_flags.go [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+	AnnotationFeatureOneAgentInitialConnectRetry    = AnnotationFeaturePrefix + "oneagent-initial-connect-retry-ms"
+	AnnotationFeatureRunOneAgentContainerPrivileged = AnnotationFeaturePrefix + "oneagent-privileged"
+	AnnotationFeatureOneAgentSecCompProfile         = AnnotationFeaturePrefix + "oneagent-seccomp-profile"
+
+	// injection (webhook).
+
+	// Deprecated: AnnotationFeatureDisableMetadataEnrichment use AnnotationFeatureMetadataEnrichment instead.
+	AnnotationFeatureDisableMetadataEnrichment = AnnotationFeaturePrefix + "disable-metadata-enrichment"
+
+	AnnotationFeatureMetadataEnrichment = AnnotationFeaturePrefix + "metadata-enrichment"
```
Comment (0sewa0) [inline pkg/api/v1beta2/dynakube/feature_flags.go:66]: same as the previous, there is a specific field, we no longer need the feature-flag

File: pkg/api/v1beta2/dynakube/feature_flags.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+	AnnotationFeatureApiRequestThreshold = AnnotationFeaturePrefix + "dynatrace-api-request-threshold"
+
+	// oneAgent.
+
+	AnnotationFeatureMultipleOsAgentsOnNode         = AnnotationFeaturePrefix + "multiple-osagents-on-node"
+	AnnotationFeatureOneAgentMaxUnavailable         = AnnotationFeaturePrefix + "oneagent-max-unavailable"
+	AnnotationFeatureOneAgentIgnoreProxy            = AnnotationFeaturePrefix + "oneagent-ignore-proxy"
+	AnnotationFeatureOneAgentInitialConnectRetry    = AnnotationFeaturePrefix + "oneagent-initial-connect-retry-ms"
+	AnnotationFeatureRunOneAgentContainerPrivileged = AnnotationFeaturePrefix + "oneagent-privileged"
+	AnnotationFeatureOneAgentSecCompProfile         = AnnotationFeaturePrefix + "oneagent-seccomp-profile"
```
Comment (0sewa0) [inline pkg/api/v1beta2/dynakube/feature_flags.go:59]: same as the previous, there is a specific field, we no longer need the feature-flag

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		}
+	}
+
+	dst.Spec.DynatraceApiRequestThreshold = src.FeatureApiRequestThreshold()
+
+	if src.FeatureOneAgentSecCompProfile() != "" {
+		dst.Spec.OneAgent.HostMonitoring = &HostInjectSpec{
+			SecCompProfile: src.FeatureOneAgentSecCompProfile(),
+		}
+	}
```
Comment (0sewa0) [inline pkg/api/v1beta2/dynakube/dynakube_conversion.go:73]: should we remove the corresponding annotations (feature-flags) from the `ObjectMeta.Annotations`?

I would say yes, so the outcome of the conversion is "cleaner".

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+			IPAddress: value.IPAddress,
+		}
+		dst.Status.OneAgent.Instances[key] = tmp
+	}
+
+	dst.Status.OneAgent.Version = src.Status.OneAgent.Version
+
+	dst.Status.Phase = src.Status.Phase
+	dst.Status.UpdatedTimestamp = src.Status.UpdatedTimestamp
+
```
Comment (0sewa0) [inline pkg/api/v1beta2/dynakube/dynakube_conversion.go:42]: related to my previous comment, we should consider using the new specific fields to add them as feature-flags here, so no information is lost during conversion.

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go (hunk 4 above)
Comment (waodim) [reply pkg/api/v1beta2/dynakube/dynakube_conversion.go:73]: do you mean to remove the annotations from the `ObjectMeta.Annotations` from v1beta1? Could you please elaborate as I am kind of lost here :sweat_smile:

File: pkg/api/v1beta2/dynakube/feature_flags.go (hunk 2 above)
Comment (waodim) [reply pkg/api/v1beta2/dynakube/feature_flags.go]: done https://github.com/Dynatrace/dynatrace-operator/pull/3092/commits/d3318cdcd1da4630154e41f06ca2c18ba5e6e48d

File: pkg/api/v1beta2/dynakube/feature_flags.go (hunk 3 above)
Comment (waodim) [reply pkg/api/v1beta2/dynakube/feature_flags.go]: done https://github.com/Dynatrace/dynatrace-operator/pull/3092/commits/d3318cdcd1da4630154e41f06ca2c18ba5e6e48d

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go (hunk 4 above)
Comment (waodim) [reply pkg/api/v1beta2/dynakube/dynakube_conversion.go]: hopefully correctly done https://github.com/Dynatrace/dynatrace-operator/pull/3092/commits/d3318cdcd1da4630154e41f06ca2c18ba5e6e48d

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go (hunk 5 above)
Comment (waodim) [reply pkg/api/v1beta2/dynakube/dynakube_conversion.go:83]: Then I think I did it correctly in https://github.com/Dynatrace/dynatrace-operator/pull/3092/commits/d3318cdcd1da4630154e41f06ca2c18ba5e6e48d

Comment (0sewa0) [reply pkg/api/v1beta2/dynakube/dynakube_conversion.go:83]: you are always creating a new map for the Annotations so only the last will take effect.
should just add the key-values, like`dst.Annotations[AnnotationFeatureOneAgentSecCompProfile] = hostMonitoring.SecCompProfile`

also be sure to take care of the `dst.Annotations == nil` scenario 😉

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	dst.Spec.NetworkZone = src.Spec.NetworkZone
+	dst.Spec.EnableIstio = src.Spec.EnableIstio
+
+	if src.Annotations == nil {
+		src.Annotations = map[string]string{}
+	}
+
+	if !dst.Spec.MetaDataEnrichment.Enabled {
+		src.Annotations[dynakube.AnnotationFeatureMetadataEnrichment] = "false"
+	}
```
Comment (0sewa0) [inline pkg/api/v1beta2/dynakube/dynakube_conversion.go:111]: I think you did a mix up here, the `v1beta2 => v1beta1` looks good, but here you are creating the `dst` so checking it makes little sense

Comment (0sewa0) [reply pkg/api/v1beta2/dynakube/dynakube_conversion.go:111]: The whole `v1beta1 => v1beta2` looks a bit like copy-pasta from the `v1beta2 => v1beta1`, and while certain actions are the same, the more complicated ones are not, here you should be removing the annotations and not removing them

**Review by chrismuellner — CHANGES_REQUESTED** [review]
Just checking occurrences of `v1beta1`/`v1alpha1` in the code base, I think we need to adapt a few more files:
- `PROJECT` has `v1alpha1` and `v1beta1` only
- `troubleshoot` package and command still use `v1beta1`
- `v1alpha1` is still used in a bunch of places (can we completely remove it?)

**Comment by waodim** [comment]
> Just checking occurrences of `v1beta1`/`v1alpha1` in the code base, I think we need to adapt a few more files:
> 
>     * `PROJECT` has `v1alpha1` and `v1beta1` only
> 
>     * `troubleshoot` package and command still use `v1beta1`
> 
>     * `v1alpha1` is still used in a bunch of places (can we completely remove it?)

- I updated `PROJECT` locally and will push soon
- In my opinion troubleshoot using `v1beta1` is okay as it is our `hub` version now, or am I wrong there?
- I did not find usages of `v1alpha1` version of the `dynakube` to be honest. When it comes to the whole `v1alpha1` package we should not remove it in my opinion because of edgeconnect

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go (hunk 6 above)
Comment (waodim) [reply pkg/api/v1beta2/dynakube/dynakube_conversion.go]: should be done https://github.com/Dynatrace/dynatrace-operator/pull/3092/commits/4dba19d3d27b6550647c1382a1ddd444947aae64

**Comment by chrismuellner** [comment]
I would assume that all our logic should use the same, latest version of `v1beta2`.

I missed that the `v1alpha1` package is required by edgeconnect, good point 😉

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+		src.Annotations = map[string]string{}
+	}
+
+	if src.Annotations[dynakube.AnnotationFeatureMetadataEnrichment] == "true" {
+		dst.Spec.MetaDataEnrichment = MetaDataEnrichment{
+			Enabled: false,
+		}
+	}
+
+	src.Annotations[dynakube.AnnotationFeatureApiRequestThreshold] = strconv.FormatInt(int64(dst.Spec.DynatraceApiRequestThreshold), 10)
```
Comment (0sewa0) [inline pkg/api/v1beta2/dynakube/dynakube_conversion.go:115]: you want to convert the feature flag into the dedicated field, and not the other way
also don't forget to remove the feature flag afterward
```
delete(dst.Annotations, dynakube.AnnotationFeatureApiRequestThreshold)
```

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+	if src.Annotations[dynakube.AnnotationFeatureOneAgentSecCompProfile] != "" {
+		secCompProfile := src.Annotations[dynakube.AnnotationFeatureOneAgentSecCompProfile]
+		dst.Spec.OneAgent.HostMonitoring.SecCompProfile = secCompProfile
+	}
```
Comment (0sewa0) [inline pkg/api/v1beta2/dynakube/dynakube_conversion.go:120]:
```suggestion
	if src.Annotations[dynakube.AnnotationFeatureOneAgentSecCompProfile] != "" {
		secCompProfile := src.Annotations[dynakube.AnnotationFeatureOneAgentSecCompProfile]
		dst.Spec.OneAgent.HostMonitoring.SecCompProfile = secCompProfile
		delete(dst.Annotations, dynakube.AnnotationFeatureOneAgentSecCompProfile)
	}
```

should remove the "moved" feature-flag, so they don't get out of sync (so its nice and clean)

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go (hunk 8 above)
Comment (waodim) [reply pkg/api/v1beta2/dynakube/dynakube_conversion.go]: should be done in https://github.com/Dynatrace/dynatrace-operator/pull/3092/commits/71219b23d3ca62cab9944fc42f986e38da78c886

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go (hunk 7 above)
Comment (waodim) [reply pkg/api/v1beta2/dynakube/dynakube_conversion.go]: should be done in https://github.com/Dynatrace/dynatrace-operator/pull/3092/commits/71219b23d3ca62cab9944fc42f986e38da78c886

File: pkg/api/v1beta2/dynakube/dynakube_conversion.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+		delete(dst.Annotations, dynakube.AnnotationFeatureMetadataEnrichment)
+	} else {
+		dst.Spec.MetaDataEnrichment = MetaDataEnrichment{
+			Enabled: address.Of(true),
+		}
+		delete(dst.Annotations, dynakube.AnnotationFeatureMetadataEnrichment)
+	}
+
+	if src.Annotations[dynakube.AnnotationFeatureApiRequestThreshold] != "" {
+		duration, err := time.ParseDuration(src.Annotations[dynakube.AnnotationFeatureApiRequestThreshold])
```
[the remark this answers is not in the document]
Comment (0sewa0) [reply pkg/api/v1beta2/dynakube/dynakube_conversion.go:124]: my attempts so far: https://github.com/Dynatrace/dynatrace-operator/commit/19907966bfc4c60346a20b772da1dd76299cd37f

[the remark this answers is not in the document]
Comment (0sewa0) [reply pkg/api/v1beta2/dynakube/dynakube_conversion.go:124]: > So basically you switched hub version
that was something I tried, it does not fix the problem

(I think its better to have the new version as the hub, if we are only using that in code, it should result in less conversions)

**Comment by 0sewa0** [comment]
Rebased PR #3188 is already merged

**Pull request closed without merging**